test/hotspot/gtest/memory/test_arena.cpp

```diff
@@ -0,0 +1,413 @@
+  if (p == NULL || s == 0) {
+    return true;
+  }
+
+  const char* first_wrong = NULL;
+  char* p2 = (char*)p;
+  const char* const end = p2 + s;
+  while (p2 < end) {
+    if (*p2 != (char)expected) {
+      first_wrong = p2;
+      break;
+    }
+    p2 ++;
+  }
+
+  if (first_wrong != NULL) {
+    tty->print_cr("wrong pattern around " PTR_FORMAT, p2i(first_wrong));
+    os::print_hex_dump(tty, (address)(align_down(p2, 0x10) - 0x10),
+                            (address)(align_up(end, 0x10) + 0x10), 1);
+  }
+
+  return first_wrong == NULL;
+}
+
+// We use this to fill the allocated ranges with a pattern to test for overwriters later.
+static void mark_range(void* p, size_t s, int mark) {
+  const char m = (char)((mark + 1) % 10); // valid marks are [1...10]
+  if (p != NULL && s > 0) {
+    ::memset(p, m, s);
+  }
+}
+
+// Check a range marked with mark_range()
+static bool check_marked_range(const void* p, size_t s, int expected_mark) {
+  const char m = (char)((expected_mark + 1) % 10); // valid marks are [1...10]
+  return check_range(p, s, m);
+}
+
+#define ASSERT_RANGE_IS_MARKED(p, size, mark) ASSERT_TRUE(check_marked_range(p, size, mark))
+static const int some_random_mark = 4711;
+
+// Helper to check the arena alignment gap pattern.
+// In debug, if arena allocation causes gaps due to alignment, a pattern is written into that gap:
+//
+// |------------------|----|-------
+// | first            |'G' | second
+// | allocation       |    | allocation
+// |------------------|----|-------
+//
+// This function checks that pattern.
+#ifdef ASSERT
+static bool check_alignment_gap_pattern(const void* p1, size_t first_allocation_size, const void* p2) {
+
+  // Omit the test if one of them is NULL
+  if (p1 == NULL || p2 == NULL) {
+    return true;
+  }
+
+  // Omit the test if it looks like the pointers are completely unrelated - p2 may live in a new Chunk.
+  if ((p2 > p1 && pointer_delta(p2, p1, 1) >= sizeof(Chunk)) ||
+      (p1 > p2 && pointer_delta(p1, p2, 1) >= sizeof(Chunk))) {
+    return true;
+  }
+
+  const char GAP_PATTERN = 'G'; // see arena.hpp
+  const char* gap = ((const char*)p1) + first_allocation_size;
+  assert(gap <= p2, "sanity");
+  const size_t len = ((const char*)p2) - gap;
+
+  if (len > 0) {
+    if (check_range(gap, len, GAP_PATTERN) == false) {
+      return false;
+    }
+    // Test the first byte after the gap too: should *not* be "G" since mark_range() does not use that pattern
+    if (gap[len] == GAP_PATTERN) {
+      return false;
+    }
+  }
+
+  return true;
+}
+#define ASSERT_GAP_PATTERN(p1, size1, p2) ASSERT_TRUE(check_alignment_gap_pattern(p1, size1, p2));
+#else
+#define ASSERT_GAP_PATTERN(p1, size1, p2)
+#endif
+
+
+// Note: any tests making assumptions about the placement of allocations needs to
+// be guarded from running with UseMallocOnly.
+
+TEST_VM(Arena, alloc_alignment_32) {
+  Arena ar(mtTest);
+  void* p1 = ar.Amalloc(1);   // just one byte, leaves _hwm unaligned at offset 1
```

Comment:
No I don't want this.  Amalloc should still round up the allocation size to 64 bits, then return the pointer aligned to 64 bits.
The test that I was going to write is 
p1  = AmallocWords(4)
p2 = Amalloc(some size that gets aligned to 64 bits)
verify p2 is 64 bit aligned.

I don't want _hwm to ever be unaligned.